### PR TITLE
feat: add getParameter api

### PR DESCRIPTION
**Pull request by ira-gordin-sap**
It was tested manually and works. I need to write 2 more tests after I'll find how to mock the '@sap/plugin' module.
I'll really appreciate you can start a review meanwhile, because the end of the sprint is coming. 
Completed including manual and unit tests.

### CODE REVIEW

**Comment by bd82** [comment]
please don't do any more force pushes as it makes it harder to do incremental review

**Comment by ira-gordin-sap** [comment]
sorry, I was sure you did not start - I only removed unused comment

File: sample-action-client/src/extension.ts [hunk 1]
```diff
@@ -20,6 +20,18 @@ export async function activate(context: vscode.ExtensionContext) {
         vscode.window.showInformationMessage(`Action ${action.name} was scheduled to run`,);
         basAPI.actions.performAction(action, {schedule: true});
     }));
+
+	context.subscriptions.push(vscode.commands.registerCommand("get.parameter", async () => {
+        const parameterName = await vscode.window.showInputBox({ prompt: "Enter Parameter Name", ignoreFocusOut: true });
+        if (!parameterName) {
```
Comment (bd82) [inline sample-action-client/src/extension.ts:26]: it is better to avoid automatic boolean conversions in JavaScript
as it can lead to strange bugs in edge cases.

Something like:
```javascript
// undefined on escape and empty on enter
if (parameterName === undefined || parameterName === "") { /* ... */} 
```

File: sample-action-client/src/extension.ts [hunk 2]
```diff
@@ -20,6 +20,18 @@ export async function activate(context: vscode.ExtensionContext) {
...
     }));
+
+	context.subscriptions.push(vscode.commands.registerCommand("get.parameter", async () => {
+        const parameterName = await vscode.window.showInputBox({ prompt: "Enter Parameter Name", ignoreFocusOut: true });
+        if (!parameterName) { 
+            // undefined in escape and empty on enter
+            return;
+        }        
+        
+        const parameterValue = await basAPI.getParameter(parameterName as string);
```
Comment (bd82) [inline sample-action-client/src/extension.ts:31]: I think the explicit casting here is not "at the correct spot".

Option1:
- add the casting where you defined `parameterName` (what is the return type of `showInputBox` ?)

Option2:
- extract the `if (!parameterName) {` block to a helper function with a **type guard** 
- https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards

we can discuss which option is better offline

Comment (ira-gordin-sap) [reply sample-action-client/src/extension.ts:31]: I'll remove the casting. I don't really need it now. Don't remember why I added it.

File: src/api.ts [hunk 3]
```diff
@@ -2,7 +2,9 @@ import * as vscode from 'vscode';
 import { performAction } from './actions/client';
 import { ActionsController } from './actions/controller';
 import { ExecuteAction, SnippetAction, CommandAction, FileAction } from './actions/impl';
+import { getLogger } from './logger/logger';
 export * from "./actions/interfaces";
+import _ = require('lodash');
```
Comment (bd82) [inline src/api.ts:7]: I recommend you only import the exact functions you need from lodash
You can use an ESM compatible lodash build from here: https://www.npmjs.com/package/lodash-es

Comment (bd82) [inline src/api.ts:7]: import order:
- don't mix import and exports. (export after all import)
- import of 3rd party libraries should appear **before** import of relative paths modules.

File: tests/api.spec.ts [hunk 4]
```diff
@@ -126,3 +175,12 @@ function promiseWithTimeout(promise: any, timeout: number){
...
 }
+
+/**
+ * Stub key-> value object to the environment variable for testing purpose
+ * @param newValues - The key-> valye to set the environment
+ */
+ function stubEnv(newValues: Record<string, unknown>): void {
+    const extendedEnv = extend({}, process.env, newValues);
+    stub(process, "env").value(extendedEnv);
+  }
```
Comment (bd82) [inline tests/api.spec.ts:186]: missing newline

File: src/api.ts [hunk 5]
```diff
@@ -35,6 +37,32 @@ export const bas = {
         return ActionsController.getAction(actionId);
     },
 
+    async getParameter(parameterName: string) : Promise<string | null> {
```
Comment (bd82) [inline src/api.ts:40]: why is it `string | null` is it not more appropriate to return `string | undefined` 

WDYT?

File: src/api.ts [hunk 6]
```diff
@@ -35,6 +37,32 @@ export const bas = {
...
+        const onBasEnv = "env.WS_BASE_URL";
```
Comment (bd82) [inline src/api.ts:41]: `"env.WS_BASE_URL";` to constants instead of magic string value?

File: src/api.ts (hunk 3 above)
Comment (ira-gordin-sap) [reply src/api.ts:7]: import {isEmpty, get} from 'lodash';
?
(We have _ in all our projects..)

File: src/api.ts (hunk 6 above)
Comment (bd82) [reply src/api.ts:41]: ignore this, see next commt

File: src/api.ts (hunk 5 above)
Comment (ira-gordin-sap) [reply src/api.ts:40]: @idoprz asked it

File: src/api.ts [hunk 7]
```diff
@@ -35,6 +37,32 @@ export const bas = {
...
+        const isInBAS = !_.isEmpty(_.get(process, onBasEnv));
```
Comment (bd82) [inline src/api.ts:42]: try using optional chaining here instead. of the `!_.isEmpty(_.get`
- https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html

File: src/api.ts (hunk 6 above)
Comment (ira-gordin-sap) [reply src/api.ts:41]: This is constant.

File: src/api.ts [hunk 8]
```diff
@@ -35,6 +37,32 @@ export const bas = {
...
+        if (!isInBAS) {
+            getLogger().trace("Running in VS Code, so returning null.");
+            return null;
+        }
+        getLogger().trace("Running on BAS.");
+
+        let sap;
+        try {
+            sap = require('@sap/plugin');
```
Comment (bd82) [inline src/api.ts:51]: try to use `https://www.npmjs.com/package/optional-require`

File: src/api.ts [hunk 9]
```diff
@@ -35,6 +37,32 @@ export const bas = {
...
+        } catch (error) {
+            getLogger().error("Failed to load @sap/plugin, so returning null.", {error});
+            return null;
+        }
+        getLogger().trace("@sap/plugin successfully loaded.");
+
+        const configuration = await sap.window.configuration();
+        getLogger().trace("Configuration successfully received.", {configuration});
+
+        const parameterValue = _.get(configuration, parameterName, null);
```
Comment (bd82) [inline src/api.ts:61]: should we return `null` or `undefined` ?
you ay be able to use optional chaining instead of `_.get`

**Comment by bd82** [comment]
Is there documentation for this `getParameter`? who is the consumer of this API?
How Would they know how to use it / in what scenarios it is relevant?

File: src/api.ts (hunk 3 above)
Comment (bd82) [reply src/api.ts:7]: yes, but that was before ESM imports
`import {foo, bar, bamba} from "../blah"`

File: src/api.ts (hunk 5 above)
Comment (bd82) [reply src/api.ts:40]: @idoprz WDYT null or undefined?

File: src/api.ts [hunk 10]
```diff
@@ -2,7 +2,9 @@ import * as vscode from 'vscode';
 import { performAction } from './actions/client';
 import { ActionsController } from './actions/controller';
 import { ExecuteAction, SnippetAction, CommandAction, FileAction } from './actions/impl';
+import { getLogger } from './logger/logger';
 export * from "./actions/interfaces";
+import _ = require('lodash');
 
 export const bas = {
```
Comment (bd82) [inline src/api.ts:9]: 1. is the type of `bas` defined anywhere?
2. evaluate separating the bas structure and the implementation of each method.
    basically what we have now does not scale, if we will have 20 methods/functions this api.ts file will be 1,000+ line long.

**Comment by ira-gordin-sap** [comment]
Yes, it should be documented. Where is a good place to document it?

File: src/api.ts (hunk 10 above)
Comment (ira-gordin-sap) [reply src/api.ts:9]: to move my api to separate file? (thought about it) to create a folder for all apis?

File: src/api.ts (hunk 8 above)
Comment (ira-gordin-sap) [reply src/api.ts:51]: When I use optional-require, I don't need to test isInBAS, right? Are you ok I'll remove it?

Comment (bd82) [reply src/api.ts:51]: not sure, lets sync offline

File: src/api.ts (hunk 10 above)
Comment (bd82) [reply src/api.ts:9]: folder for all APIs seems like a good idea.
But start with only your new API (**don't** make this PR into a re-factor)

**Comment by bd82** [comment]
Yes, it should be documented. Where is a good place to document it?

There is a `types` folder in the root of this repo, I am not sure what it currently contains.
We may need to handle the documentation question separately outside this PR.

File: src/api.ts (hunk 5 above)
Comment (ira-gordin-sap) [reply src/api.ts:40]: This is from @idoprz 
If the user tries to get value of key that does not exist it returns null. If running in VSCode all keys will return null as value for now.

File: src/api.ts (hunk 9 above)
Comment (ira-gordin-sap) [reply src/api.ts:61]: Works perfect for undefined "const parameterValue = configuration?.[parameterName];", but not for null. Or you know a better way?

File: src/api.ts (hunk 5 above)
Comment (ira-gordin-sap) [reply src/api.ts:40]: I talked to Ido and it doesn't matter for him null or undefined. 
I do love null, because undefined can come from some bug and null is more sure, but can change if you are sure it is better.

File: src/api.ts [hunk 11]
```diff
@@ -35,6 +37,10 @@ export const bas = {
         return ActionsController.getAction(actionId);
     },
 
+    async getParameter(parameterName: string) : Promise<string | undefined> {
```
Comment (bd82) [inline src/api.ts:40]: `  getParameter : getParameter`

File: src/apis/parameters.ts [hunk 12]
```diff
@@ -0,0 +1,29 @@
+//import {isEmpty, get} from 'lodash';
+import { getLogger } from '../logger/logger';
+
+export async function getParameter(parameterName: string) : Promise<string | undefined> {
+    // const onBasEnv = "env.WS_BASE_URL";
```
Comment (bd82) [inline src/apis/parameters.ts:5]: delete commnets...

File: src/apis/parameters.ts [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+    //const isInBAS = get(process, onBasEnv)?.is;
+    // if (!isInBAS) {
+    //     getLogger().trace("Running in VS Code, so returning undefined.");
+    //     return undefined;
+    // }
+    // getLogger().trace("Running on BAS.");
+
+    const optionalRequire = require("optional-require")(require);
+    const noSapPlugin = "noSapPlugin";
+    const sapPlugin = optionalRequire('@sap/plugin') || noSapPlugin;
```
Comment (bd82) [inline src/apis/parameters.ts:16]:
```javascript
  const noSapPlugin = "NO_SAP_PLUGIN_FOUND";
  const sapPlugin = optionalRequire('@sap/plugin') ?? noSapPlugin;
```

File: tests/api.spec.ts [hunk 14]
```diff
@@ -16,6 +16,7 @@ const testVscode = {
 mockVscode(testVscode, "src/api.ts");
 
 import { bas } from "../src/api";
+import * as parameters from "../src/apis/parameters";
```
Comment (bd82) [inline tests/api.spec.ts:19]: probably don't need this import with better BAS object definition

File: tests/api.spec.ts [hunk 15]
```diff
@@ -114,6 +111,14 @@ describe("api unit test", () => {
         extensionsMock.expects("getExtension").withExactArgs("myExt").returns(undefined);
         await expect(bas.getExtensionAPI("myExt")).to.be.rejectedWith(`Extension myExt is not loaded`);
     });
+
+    it("getParameter", async() => {
```
Comment (bd82) [inline tests/api.spec.ts:115]: this test will no longer be needed

File: tests/apis/parameters.spec.ts [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+import { extend } from "lodash";
+import { createSandbox, stub } from "sinon";
+import { getParameter } from '../../src/apis/parameters';
+
+describe("getParameter", () => {
+    let sandbox: any;
+    const WS_BASE_URL = "WS-BASE-VALUE";
+    const parameterName = "param1";
+
+    before(() => {
```
Comment (bd82) [inline tests/apis/parameters.spec.ts:11]: is the sandbox required?

File: tests/apis/parameters.spec.ts [hunk 17]
```diff
@@ -0,0 +1,139 @@
+import { expect } from "chai";
+import { extend } from "lodash";
+import { createSandbox, stub } from "sinon";
+import { getParameter } from '../../src/apis/parameters';
+
+describe("getParameter", () => {
```
Comment (bd82) [inline tests/apis/parameters.spec.ts:6]: https://mochajs.org/#bdd (refactor to this style)

File: tests/apis/parameters.spec.ts [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+    //     expect(parameterValue).to.be.undefined;
+    // });
+
+    it("no @sap/plugin loaded ---> returns undefined", async () => {
+        stubEnv({ WS_BASE_URL: WS_BASE_URL });
+        const parameterValue = await getParameter(parameterName);
+        expect(parameterValue).to.be.undefined;
+    });
+
+    it("@sap/plugin loaded, configuration is null ---> returns undefined", async () => {
```
Comment (bd82) [inline tests/apis/parameters.spec.ts:31]:
```javascript
context("@sap/plugin loaded", () => {
  describe("when configuration is null", () => {
    let requireMock
    
    before(() => {
      requireMock = require('mock-require');
      const sapPlugin = {
        window: {
          configuration: () => null
        }
      };
      requireMock('@sap/plugin', sapPlugin);
    })

    it("returns undefined", async () => {
      const parameterValue = await getParameter(parameterName);
      expect(parameterValue).to.be.undefined;
    });
    
    after(() => {
      requireMock.stop('@sap/plugin');
    })
  })
})
```

File: tests/apis/parameters.spec.ts [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+        requireMock.stop('@sap/plugin');
+    });
+
+    it("@sap/plugin loaded, configuration is undefined ---> returns undefined", async () => {
+        stubEnv({ WS_BASE_URL: WS_BASE_URL });
+        const requireMock = require('mock-require');
+
+        const sapPlugin = {
+            window: {
+                configuration: () => undefined
```
Comment (bd82) [inline tests/apis/parameters.spec.ts:53]: inspect what is possible according to `window.configuration()` API.
and consider which edge cases need to be checked.

File: tests/apis/parameters.spec.ts [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+            }
+        };            
+        requireMock('@sap/plugin', sapPlugin);
+
+        const parameterValue = await getParameter(parameterName);
+        expect(parameterValue).to.be.equal(expectedParameterValue);
+        requireMock.stop('@sap/plugin');
+    });
+
+    it("@sap/plugin loaded, parameter name exists in the configuration with undefined value ---> returns the parameter value", async () => {
```
Comment (bd82) [inline tests/apis/parameters.spec.ts:114]: like before you may not need to check all edge cases of parameters values (null/undefined) and instead can rely on member access syntax `["foo"]`

File: tests/apis/parameters.spec.ts (hunk 19 above)
Comment (ira-gordin-sap) [reply tests/apis/parameters.spec.ts:53]: The API:
  export namespace window {
    export function configuration(): Promise<{ [id: string]: string }>;
  }
They has initialization to empty in the code:
const configuration: { [id: string]: string } = {};
So currently it can be only empty and not null and not undefined.
? behaves the same on null and undefined
So I am keeping only null configuration.

File: tests/apis/parameters.spec.ts [hunk 21]
```diff
@@ -0,0 +1,94 @@
+import * as vscode from "vscode";
```
Comment (bd82) [inline tests/apis/parameters.spec.ts:1]: i think this is import is unused and will cause runtime errors during tests

File: tests/apis/parameters.spec.ts [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+            });
+            
+            after(() => {
+                requireMock.stop('@sap/plugin');
+            });
+        });
+    });
+      
+    context("@sap/plugin loaded", () => {
+        // no test for configuration containing other parameters, because [] behaves the same
```
Comment (bd82) [inline tests/apis/parameters.spec.ts:42]: [] --> to 'member access'

because `[]` can also reference an array or destructuring in JavaScript syntax.

**Review by bd82 — APPROVED** [review]

File: .mocharc.yaml [hunk 23]
```diff
@@ -0,0 +1,6 @@
+require:
```
Comment (bd82) [inline .mocharc.yaml:1]: use a `.js` file instead of yaml for .mocharc
as it would allow better re-use / combining

File: package.json [hunk 24]
```diff
@@ -56,14 +56,15 @@
 		"compile": "tsc -p ./",
 		"watch": "tsc -watch -p ./",
 		"pretest": "npm run compile",
-		"test": "nyc mocha -p tsconfig.json --recursive --no-parallel tests/**/*.spec.ts",
+		"test": "nyc mocha -p tsconfig.json --config ./.mocharc.yaml --no-parallel",
```
Comment (bd82) [inline package.json:59]: I think the mocharc file should be automatically loaded if it is in the same folder

**Review by bd82 — APPROVED** [review]
you can fix the mocha configuration in a different PR

**Review by slavik-lvovsky — APPROVED** [review]

File: .mocharc.yaml (hunk 23 above)
Comment (ira-gordin-sap) [reply .mocharc.yaml:1]: this is .js now
